core/src/main/scala/kafka/log/LogCleaner.scala

```diff
@@ -701,11 +719,17 @@ private[log] class Cleaner(val id: Int,
       // if any messages are to be retained, write them out
       val outputBuffer = result.outputBuffer
       if (outputBuffer.position() > 0) {
+        if (destSegment.isEmpty) {
+          // create a new segment with a suffix appended to the name of the log and indexes
+          destSegment = Some(LogCleaner.createNewCleanedSegment(log, result.minOffset()))
+          transactionMetadata.cleanedIndex = Some(destSegment.get.txnIndex)
```

Comment:
Re: changing transaction code
Since cleanedIndex is not created until here, we need to delay appends. Originally it was created at the start of `cleanSegments`. In that method, we can not guarantee the log has been created until `cleanInto` returns with a defined `cleanedSegment`. In `cleanSegments` there is a chance to call `onControlBatchRead` which previously just appeneded transactions, but can no longer do so.

Reply:
It may be possible to call `transactionMetadata.appendTransactionIndex()` here though to append a little bit sooner.

Reply:
Another option is to check if the `cleanedIndex` exists and append immediately if it does, otherwise put it in `toAppend`. Then we apply the changes only when the `cleanedIndex` is first defined.

Reply:
Ok, I think I understand. One thing that makes this a little confusing is the need to reverse the collection. Why don't just build the index in order?

I am debating whether this is good enough. A potential problem is that we don't have a guarantee on the size of the index. In the common case, it should be small, but there is nothing to prevent an entire segment from being full of aborted transaction markers. Currently we wait until `filterInto` returns before creating the new segment, but maybe instead we could do it in `checkBatchRetention` after the first retained batch is created?

Reply:
Oh I think the reverse is just because of how I append the transactions to the front of the list instead of the end. There's a better way to do that, so I'll fix it.

Reply:
Locally I have it filterTo returns. I'll look for a better way to do it.

Reply:
I can do it in `onControlBatchRead` directly I believe. I'll add to toAppend when there is no cleanedIndex and as soon as there is, I'll add those transactions + any new ones read. From then on, I can just add the transactions in `onControlBatchRead`

Reply:
Sorry to keep pressing this, but do you have an update here? It would be nice to keep the behavior here on par with what is upstream. Really if you think about the way this is working, we could write to the cleaned segment directly from `filterTo` instead of building a buffer in memory, but we can leave that as a follow-up improvement.